src/main/scala/Decoder.scala

```diff
@@ -26,116 +25,112 @@ class Decoder(bitWidth: Int = 32) extends Module {
         // Stores
-        BitPat("b?????????????????000?????0100011")  -> List(INST_S,      SB, false.B,   false.B, true.B,    false.B, false.B,   true.B),
-        BitPat("b?????????????????001?????0100011")  -> List(INST_S,      SH, false.B,   false.B, true.B,    false.B, false.B,   true.B),
-        BitPat("b?????????????????010?????0100011")  -> List(INST_S,      SW, false.B,   false.B, true.B,    false.B, false.B,   true.B),
+        BitPat("b?????????????????000?????0100011")  -> BitPat(Cat(INST_S.asUInt,       SB.asUInt, false.B,   false.B, true.B,    false.B, false.B,   true.B)),
+        BitPat("b?????????????????001?????0100011")  -> BitPat(Cat(INST_S.asUInt,       SH.asUInt, false.B,   false.B, true.B,    false.B, false.B,   true.B)),
+        BitPat("b?????????????????010?????0100011")  -> BitPat(Cat(INST_S.asUInt,       SW.asUInt, false.B,   false.B, true.B,    false.B, false.B,   true.B))
       )
-    ) // format: on
+    ,
+                                                        BitPat(Cat(IN_ERR.asUInt, ERR_INST.asUInt, false.B,  false.B, false.B,    false.B, false.B,  false.B)) // Default values
+      // format: on
+      ),
+    )
+    .asTypeOf(new DecType)
 
-  io.DecoderPort.rd       := 0.U
-  io.DecoderPort.rs1      := 0.U
-  io.DecoderPort.rs2      := 0.U
-  io.DecoderPort.imm      := 0.S
-  io.DecoderPort.inst     := signals(1)
-  io.DecoderPort.toALU    := signals(2)
-  io.DecoderPort.branch   := signals(3)
-  io.DecoderPort.use_imm  := signals(4)
-  io.DecoderPort.jump     := signals(5)
-  io.DecoderPort.is_load  := signals(6)
-  io.DecoderPort.is_store := signals(7)
-
-  switch(signals(0)) {
-    is(INST_R) {
-      io.DecoderPort.rd  := io.DecoderPort.op(11, 7)
-      io.DecoderPort.rs1 := io.DecoderPort.op(19, 15)
-      io.DecoderPort.rs2 := io.DecoderPort.op(24, 20)
-    }
-    is(INST_I) {
-      io.DecoderPort.rd  := io.DecoderPort.op(11, 7)
-      io.DecoderPort.rs1 := io.DecoderPort.op(19, 15)
-      io.DecoderPort.imm := ImmGenerator(INST_I, io.DecoderPort.op)
-    }
-    is(INST_S) {
-      io.DecoderPort.rs1 := io.DecoderPort.op(19, 15)
-      io.DecoderPort.rs2 := io.DecoderPort.op(24, 20)
-      io.DecoderPort.imm := ImmGenerator(INST_S, io.DecoderPort.op)
-    }
-    is(INST_B) {
-      io.DecoderPort.rs1 := io.DecoderPort.op(19, 15)
-      io.DecoderPort.rs2 := io.DecoderPort.op(24, 20)
-      io.DecoderPort.imm := ImmGenerator(INST_B, io.DecoderPort.op)
-    }
-    is(INST_U) {
-      io.DecoderPort.rd  := io.DecoderPort.op(11, 7)
-      io.DecoderPort.imm := ImmGenerator(INST_U, io.DecoderPort.op)
-    }
-    is(INST_J) {
-      io.DecoderPort.rd  := io.DecoderPort.op(11, 7)
-      io.DecoderPort.imm := ImmGenerator(INST_J, io.DecoderPort.op)
-    }
-  }
+  io.DecoderPort.rd  := io.DecoderPort.op(11, 7)
+  io.DecoderPort.rs1 := io.DecoderPort.op(19, 15)
+  io.DecoderPort.rs2 := io.DecoderPort.op(24, 20)
```

Comment:
If you don't try to reuse the datapath register, just do this assign is enough ;p